src/vs/vscode.proposed.d.ts

```diff
@@ -1554,24 +1561,72 @@ declare module 'vscode' {
 	}
 
 	export interface NotebookProvider {
-		resolveNotebook(editor: NotebookEditor): Promise<void>;
-		executeCell(document: NotebookDocument, cell: NotebookCell | undefined): Promise<void>;
+		resolveNotebook(editor: NotebookEditor): Promise<void>; // rchiodo: Not sure what this is for? Notebook is loaded? Like prime the server? Or does this somehow generate the document?
 		save(document: NotebookDocument): Promise<boolean>;
 	}
 
+	/**
+	 * Result of an executeCell.
+	 */
+	export interface NotebookCellExecutionResult {
+		/**
+		 * Event fired when new cell output comes through
+		 */
+		readonly onCellOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when output should be cleared. Boolean flag indicates if should wait for next onOutput for clear.
+		 */
+		readonly onClear: Event< boolean>;
+		/**
+		 * Event fired when document wide output comes through
+		 */
+		readonly onDocumentOutput: Event<CellOutput[]>;
+		/**
+		 * Event fired when execution is complete
+		 */
+		readonly onComplete: Event<void>;
+	}
+
+	/**
+	 * Allows execution of notebook cells.
+	 */
+	export interface NotebookExecution {
+		readonly executionCount: number;
+		executeCell(document: NotebookDocument, cell: NotebookCell): NotebookCellExecutionResult;
+	}
+
+	/**
+	 * Allows observing the execution of cells from a NotebookExecution
+	 */
+	export interface NotebookExecutionObserver {
+		onExecuteCell(document: NotebookDocument, cell: NotebookCell, eventProvider: NotebookCellExecutionResult): void;
```

Comment:
Not sure this is an observer, in fact it provides yet another hook to add event handlers into `eventProvider`. 

I.e. there are two or more steps:
1. registerNotebookExecutionObserver
2. eventProviders.onClear....

I would have expected the observer to have methods on it to get the information...
as follows:
```typescript
	export interface NotebookCellExecutionObserver {
		onCellOutput(document: NotebookDocument, cell: NotebookCell...)
        onCel.....
	}

Personally I like this, then there's very little API that needs to be exposed from VSCode.
Else they need to expose a lot about the structure of how things are managed and events exposed, and the like.
```

Reply:
Okay but it would need onExecute as well. You're just eliminating the extra code of signing up for the events.

Reply:
Not really (though that's a benefit), and it allows VSC to  change their API without having to worry much about the consumders
Else with the other path VSC (API publisher) will be forced to expose a specific structure that they need to adhere to..
E.g. its possible we could have separate classes that deal with execution, and another class for deleting cells (hence clear event is somewhere else), etc... 
This may not be the case, but I think it allows the publisher to evolve/change without having to worry about breaking any API...

Basically if I was exposing an API, I'd allow for my (publisher) end to have most flexibility..